### ISSUE TITLE
Reviewing Assignment

### ISSUE THREAD

**Issue by rayanirban** [dlmbl/boot#9]
Request to review the notebooks in the main branch inside a colab environment.

**Comment by afoix** [comment]
# Review until Checkpoint 1

## Chapter 0
- [x] At the beginning of chapter 0 you introduce three very important libraries such as `pathlib, urllib.request, zipfile and tqdm` . Perhaps it would be interesting to explain in 4-5 words what is for example `tqdm`. In case they don't have experience in coding.
- [x] In task 0.2, you use the sentence: "Can you a bash command to programmatically count the number of images and masks present in the download/images directory." I believe is missing a question mark and a verb such as "use", "run" or some synonyms. Otherwise you have to write "You can use a bash command...".

## Chapter 1
- [x] Task 1.1. same: you wrote: "Can you visualize the corresponding mask for the image above." either you add a question mark at the end or you rephrase the sentence saying something on the lines of "You can visualize...".
- [x] Task 1.1. in the Function description section you have 2 typos: you wrote "_Funtion description_" and it should be spelled "**Function** description". Also,  you wrote "_There is a way inb python to get the description of a function._" I believe the b is a typo that should be removed.
- [x] You wrote "_It often helps model training, if we provide image inputs to the model which are between [0,1] intensities._". I believe a good rephrasing of the sentence would be: "Providing image inputs with intensities between [0, 1] often helps when training the model. ".
- [x] You wrote: "_This is because the gradients are more stable and the model converges faster. And the model is not biased towards any particular intensity values._" and I would write same with a coma before the "and", and write the "and" with lower case. 
- [x] At the end of checkpoint 1, you wrote "This is useful if there are" and stop the sentence. It feels that there is something missing there 😄 


I will continue later/tomorrow with the review of the next Chapters. Great job @rayanirban 🥳

**Comment by afoix** [comment]
# Review until check point 5

## Chapter 2
- [x] I think Task 2.1. is a bit ambiguous, what do you mean by "left portion"? Which % of the images is this? How much they need to show?
- [x] In the part of downsampling I would develop a little bit more: Why is it OK downsampling? Is there any risk? Is that a "lossy" process?
- [x] In task 2.2. same rephrasing as suggested before: "Can you see that the image on the right lacks some detail on account of being downsampled." is this a question or a explanation?

## Chapter 3
- [x] In task 3.2. you say "assume that your image is square and that your filter is square and has an odd width." I think it will be interesting to develop a little bit more: what is a filter? Why we use filters? Why it needs to be square in the case you specify and why it needs to have an odd width?
- [x] I saw the TQDM explanation now 😄 I'd put it the first time you use the library, it would be useful for the students to understand what are they using. 
- [x] In task 3.4 when you introduce the Sobel filter, please put a [reference](https://en.wikipedia.org/wiki/Sobel_operator) for more information 😃

## Chapter 4
- [x]  In the introduction of chapter 4 you use the word "inflate", I'd suggest to use: increase or make it bigger. 
- [x] I will put a reference to [Torchvision](https://pytorch.org/vision/stable/index.html) and [Monai](https://monai.io/) in the introduction so if they are interested they can read later on about them. 
- [x] Maybe worth to ask about disadvantages of augmentation and why we need to be careful. 


Excellent job @rayanirban I am looking forward to help you with the exercise! i am sure the students will learn a lot and have a lot of fun!

**Comment by edyoshikun** [comment]
Here is my review of the notebook. I tested this by running the `solution.py` and the `solution.ipynb`. 

## Chapter 0
- [x] For this notebook, I didn't see an environment or in-line `pip installing` of some packages that are used. 
- [x] Task 0.2. Is the expectation that they know shell commands (i.e wc, ls -l , etc)? Perhaps point them to a [cheet sheet]. No need to explain each command but a resource should be good enough (https://linuxstans.com/wp-content/uploads/2023/06/bash-cheat-sheet.png)

## Chapter 1
- [x] I often times I find myself having to add/remove dimensions. (i.e `squeeze(), unsqueeze(), arr[np.newaxis,...]`,etc). Perhaps this can be added after `Task 1.5`
- [x] This one is super minor. Just highlighting that on Task 1.8 you could use `tqdm` in that simple for loop to demonstrate it earlier in the notebook. It's a single loop vs two loops. 
- [x] I assume in this notebook we aren't using anything fancy to look at the masks (i.e `skimage.color.label2rgb` is a cool one). It would be nice to show that the instance segmentation values map to some ID number. It could be as simple as adding the viridis colorbar.

## Chapter 2
- [x] I agree with Anna. Probably rephase to let's look at crop of the image instead of using `portion`

## Chapter 3
- [x] Minor typo right above Task 3.2 `Here we will optionally demosntrate TQDM..` 
- [x] For most of the tasks here, I think if you used a crop of the image it would be better. It's hard to see how the filters are doing anything don't the whole FOV. It will also be faster to iterate/play with different kernels.

## Chapter 4
- [x] Task 4.1. I was testing this on the `solution.py` file, so I didn't have `torchvision`. Not sure if the google collab has it pre-installed. This refers back to my first comment in chapter 0

## Chapter 5 (optional)
- [x] Maybe the mapping of masks to values by looking at the colorbar from Chapter 1 can go here instead. The only downside is that this is optional.. 


Great job @rayanirban and see you soon!

**Comment by Ben-Salmon** [comment]
# Review

Hi @rayanirban, the notebook looks great. The students should get a good understanding on the foundations of handling data. The pacing of the tasks is really good too. There are just a few things I would change:

## Welcome

- [x]  I think "Data Augmentation (This is important as it helps in artificially increasing the size of the training data which is useful for training models in a memory efficient way)" should be changed to "...training models in a *data* efficient way," since augmentation doesn't change the memory requirements for training.

## Chapter 1

- [x] Task 1.2: The text is duplicated from the cell above.
- [x] Normalizing Images: If all pixels in all channels are normalised together, e.g., by dividing them all by 255, then the stability of gradients is unaffected. This is because there has been no change in pixel intensities relative to each other. For the same reason, the model could still be biased towards particular intensity values, because some pixels could still have a much higher relative intensity. All normalisation does is reduce the scale of gradients, which allows us to use higher learning rates. If the data were unnormalized, but we reduced the learning rate accordingly, then there should be no difference in performance. Alternatively, batch normalisation would make the scale of the data irrelevant except for the first layer of the network. Normalisation is the standard because the default learning rate values in DL libraries assume the magnitude of the loss will be in the typical range for normalised inputs.
Maybe this section could instead just say that normalisation reduces the magnitude of gradients, allowing us to use higher learning rates.
- [x] Task 1.8: Instead of showing a random image-mask pair, the students could choose their own indexes, encouraging them to explore the data.
- [x] The visualize function could have a `titles` argument to make it clearer which image is which.

## Chapter 2

- [x] Downsampling: It might be worth explaining how the downsample factor is applied in the cell, that [::factor] is saying go from the start of the image to the end selecting every factor pixel.
- [x] Bonus Task for Chapter 2: Cropping does not increase the amount of data. If memory were not an issue, it would always be better or at least equivalent to train the network with full sized images rather than with crops. This is because cropping and downsampling do not introduce new examples of image-mask relationships. The only advantage to cropping and downsampling that I can think of is that they save memory, or possibly as a form of regularisation.

## Chapter 3

- [x] Task 3.1: I made the mistake of using `np.random.randint` to select batch indexes, which could give duplicate images in a batch. Students might do the same and, when they look at the solutions, not notice the difference that using `random.choice` makes. There could be a hint telling them to make should their batch contains 4 *unique* images.

See you in Woods Hole!

**Comment by rayanirban** [comment]
@afoix: Thanks! All the changes/suggestions are incorporated. 

@edyoshikun: Thanks for the feedback and suggestions. Couple of things:

> For this notebook, I didn't see an environment or in-line pip installing of some packages that are 

Nothing need to be installed

> I often times I find myself having to add/remove dimensions. (i.e squeeze(), unsqueeze(), arr[np.newaxis,...],etc). Perhaps this can be added after Task 1.5

I have added `np.newaxis`. I am not adding `squeeze` and `unsqueeze` or any other thing here as we have not introduced torch tensors yet.

> This one is super minor. Just highlighting that on Task 1.8 you could use tqdm in that simple for loop to demonstrate it earlier in the notebook. It's a single loop vs two loops.

I wanted students to focus on the filename printing and the loop is quite fast so I avoided adding `tqdm` here.

>  I assume in this notebook we aren't using anything fancy to look at the masks (i.e skimage.color.label2rgb is a cool one). It would be nice to show that the instance segmentation values map to some ID number. It could be as simple as adding the viridis colorbar.

Yes, we are not using anything fancy. I have written a sentence about this in Task 1.8 on how to interpret the colors in the mask and have added a note in the colorbar exercise in Chapter 5.

> Task 4.1. I was testing this on the solution.py file, so I didn't have torchvision. Not sure if the google collab has it pre-installed. This refers back to my first comment in chapter 0

This is not a programming task, the point is to go over to torchvision website and familiarize with the library.

> Maybe the mapping of masks to values by looking at the colorbar from Chapter 1 can go here instead. The only downside is that this is optional..

See above :)

@Ben-Salmon: Thanks for the feedback. One clarification:

>The visualize function could have a titles argument to make it clearer which image is which.

As we use the `visualize` function in different cases, titles was not used, I think this is fine. 

**Once again, thank you all for your feedbacks :), can't wait to see you all in Woods Hole 👯**

**All issues addressed in #10**

**Comment by rayanirban** [comment]
All issues addressed in #10
